src/hotspot/cpu/s390/c1_MacroAssembler_s390.cpp

```diff
@@ -102,40 +101,45 @@ void C1_MacroAssembler::lock_object(Register hdr, Register obj, Register disp_hd
     z_btrue(slow_case);
   }
 
-  // and mark it as unlocked.
-  z_oill(hdr, markWord::unlocked_value);
-  // Save unlocked object header into the displaced header location on the stack.
-  z_stg(hdr, Address(disp_hdr, (intptr_t)0));
-  // Test if object header is still the same (i.e. unlocked), and if so, store the
-  // displaced header address in the object header. If it is not the same, get the
-  // object header instead.
-  z_csg(hdr, disp_hdr, hdr_offset, obj);
-  // If the object header was the same, we're done.
-  branch_optimized(Assembler::bcondEqual, done);
-  // If the object header was not the same, it is now in the hdr register.
-  // => Test if it is a stack pointer into the same stack (recursive locking), i.e.:
-  //
-  // 1) (hdr & markWord::lock_mask_in_place) == 0
-  // 2) rsp <= hdr
-  // 3) hdr <= rsp + page_size
-  //
-  // These 3 tests can be done by evaluating the following expression:
-  //
-  // (hdr - Z_SP) & (~(page_size-1) | markWord::lock_mask_in_place)
-  //
-  // assuming both the stack pointer and page_size have their least
-  // significant 2 bits cleared and page_size is a power of 2
-  z_sgr(hdr, Z_SP);
-
-  load_const_optimized(Z_R0_scratch, (~(os::vm_page_size()-1) | markWord::lock_mask_in_place));
-  z_ngr(hdr, Z_R0_scratch); // AND sets CC (result eq/ne 0).
-  // For recursive locking, the result is zero. => Save it in the displaced header
-  // location (null in the displaced hdr location indicates recursive locking).
-  z_stg(hdr, Address(disp_hdr, (intptr_t)0));
-  // Otherwise we don't care about the result and handle locking via runtime call.
-  branch_optimized(Assembler::bcondNotZero, slow_case);
-  // done
-  bind(done);
+  if (LockingMode == LM_LIGHTWEIGHT) {
+    Unimplemented();
+  } else if (LockingMode == LM_LEGACY) {
+    NearLabel done;
```

Comment:
I don't understand: should "consistent handling of UseHeavyMonitors" not lead to LM_MONITOR be handled here? E.g. by directly jumping into the slow case, without bothering with stack locking?

Reply:
I was confused about this, too. But, `LM_MONITOR` is already checked in `LIR_Assembler::emit_lock` and `LIR_Assembler::emit_unwind_handler()`, so we don't reach here. x86 and aarch64 have a similar implementation. I think we could at least assert that LockingMode != LM_MONITOR.